### PR TITLE
pkcs11/_pkcs11.pyx: add wait_for_slot_event

### PR DESCRIPTION
**Pull request by grant100**
Add wait for slot event functionality to the lib class. Useful for detecting if a token was removed/added. 

I went back and forth on how to represent the CKR_NO_EVENT response from the api and whether it should be an int or raise an exception (even though it's not really an error).

### CODE REVIEW

File: pkcs11/_pkcs11.pyx [hunk 1]
```diff
@@ -1550,6 +1550,19 @@ cdef class lib:
...
+        cdef CK_FLAGS flag = 0
+        cdef CK_SLOT_ID slot_id
+
+        if not blocking:
+            flag = 1
+
+        rv = _funclist.C_WaitForSlotEvent(flag, & slot_id, NULL)
+        assertRV(rv)
+        if rv == CKR_NO_EVENT:
+            return -1
```
Comment (danni) [inline pkcs11/_pkcs11.pyx:1563]: Magic values are not the right answer. This isn't C :-P You could use a sentinel value, but truthfully an exception already fulfils that need. Just let it throw the NO_EVENT exception.

File: pkcs11/_pkcs11.pyx [hunk 2]
```diff
@@ -1550,6 +1550,19 @@ cdef class lib:
...
             return token
 
+    def wait_for_slot_event(self, blocking=False):
+        cdef CK_FLAGS flag = 0
+        cdef CK_SLOT_ID slot_id
+
+        if not blocking:
+            flag = 1
+
+        rv = _funclist.C_WaitForSlotEvent(flag, & slot_id, NULL)
```
Comment (danni) [inline pkcs11/_pkcs11.pyx:1560]: Remove extra space between & and s

Comment (danni) [reply pkcs11/_pkcs11.pyx:1560]: You also want to drop the GIL while waiting here, else we learned you won't be able to use this method with asyncio. Have a look at the `with nogil` context blocks.

File: pkcs11/_pkcs11.pyx [hunk 3]
```diff
@@ -1550,6 +1550,19 @@ cdef class lib:
         except StopIteration:
             return token
 
+    def wait_for_slot_event(self, blocking=False):
```
Comment (danni) [inline pkcs11/_pkcs11.pyx:1553]: I think the default is probably blocking?

File: pkcs11/_pkcs11.pyx [hunk 4]
```diff
@@ -1550,6 +1550,19 @@ cdef class lib:
...
+        cdef CK_FLAGS flag = 0
+        cdef CK_SLOT_ID slot_id
+
+        if not blocking:
+            flag = 1
```
Comment (danni) [inline pkcs11/_pkcs11.pyx:1558]: Use `CKF_DONT_BLOCK`

File: pkcs11/_pkcs11.pyx [hunk 5]
```diff
@@ -1550,6 +1550,19 @@ cdef class lib:
...
+
+        rv = _funclist.C_WaitForSlotEvent(flag, & slot_id, NULL)
+        assertRV(rv)
+        if rv == CKR_NO_EVENT:
+            return -1
+        return slot_id
```
Comment (danni) [inline pkcs11/_pkcs11.pyx:1564]: Let's return the actual `Slot` object?

**Review by danni — CHANGES_REQUESTED** [review]
Needs to drop the GIL for asyncio to work. Also it would be good to have a test, if we can work out how to generate slot events in SoftHSM2.

File: pkcs11/_pkcs11.pyx (hunk 5 above)
Comment (grant100) [reply pkcs11/_pkcs11.pyx:1564]: I like the idea of only returning the slot_id and giving the developer the option of scanning for the slot or not. In some cases it won't make sense e.g. If the event is for removal/detachment, there won't be a Slot to return.

File: pkcs11/_pkcs11.pyx (hunk 4 above)
Comment (grant100) [reply pkcs11/_pkcs11.pyx:1558]: Will do

File: pkcs11/_pkcs11.pyx (hunk 3 above)
Comment (grant100) [reply pkcs11/_pkcs11.pyx:1553]: I'm not sure, but I'm not picky, I'll make the change

File: pkcs11/_pkcs11.pyx (hunk 5 above)
Comment (danni) [reply pkcs11/_pkcs11.pyx:1564]: How would you know if the `slot_id` is a slot or not? It makes sense to me to do that work for them and return a slot.

File: pkcs11/_pkcs11.pyx (hunk 3 above)
Comment (danni) [reply pkcs11/_pkcs11.pyx:1553]: It's the default in PKCS#11 itself (flags = 0).

Comment (grant100) [reply pkcs11/_pkcs11.pyx:1553]: I see what you mean, the example is showing flag=0. I'll make that change.

File: pkcs11/_pkcs11_defn.pxd [hunk 6]
```diff
@@ -150,6 +150,7 @@ cdef extern from '../extern/cryptoki.h':
     cdef enum:  # CK_FLAGS
         CKF_RW_SESSION,
         CKF_SERIAL_SESSION,
+        CKF_DONT_BLOCK,
```
Comment (danni) [inline pkcs11/_pkcs11_defn.pxd:153]: Watch sorting order.

File: pkcs11/_pkcs11.pyx [hunk 7]
```diff
@@ -1550,6 +1550,18 @@ cdef class lib:
         except StopIteration:
             return token
 
+    def wait_for_slot_event(self, blocking=False):
+        cdef CK_SLOT_ID slot_id
+        cdef CK_FLAGS flag = CKF_DONT_BLOCK
```
Comment (danni) [inline pkcs11/_pkcs11.pyx:1555]: Start with flags `flags = 0`

File: pkcs11/_pkcs11.pyx [hunk 8]
```diff
@@ -1550,6 +1550,18 @@ cdef class lib:
...
+
+        if blocking:
+            flag ^= CKF_DONT_BLOCK
```
Comment (danni) [inline pkcs11/_pkcs11.pyx:1558]:
```
if not blocking:
   flags |= CKF_DONT_BLOCK
```

This makes the whole thing more lucid to read. Keeps the naming consistent with PKCS#11 etc.

File: pkcs11/_pkcs11.pyx (hunk 5 above)
Comment (grant100) [reply pkcs11/_pkcs11.pyx:1563]: How would you feel about returning all available slots?

Comment (danni) [reply pkcs11/_pkcs11.pyx:1563]: I don't know what we'd do with all slots. Is there a reason we can't return the slot with that id?

Comment (grant100) [reply pkcs11/_pkcs11.pyx:1563]: No I guess not, I went ahead an added that.

**Review by danni — APPROVED** [review]

**Pull request merged**